File: app/src/test/java/com/papi/nova/manager/FeatureFlagManagerTest.java
package com.papi.nova.manager;

import com.papi.nova.api.PolarisCapabilities;

import org.junit.After;
import org.junit.Test;

import java.lang.reflect.Field;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

public class FeatureFlagManagerTest {

    @After
    public void tearDown() {
        FeatureFlagManager.INSTANCE.reset();
    }

    @Test
    public void hasCursorVisibilityControl_reflectsCapabilities() throws Exception {
        setCapabilities(new PolarisCapabilities(
                "polaris",
                "1.0.0",
                new PolarisCapabilities.Features(false, false, false, false, false, true),
                new PolarisCapabilities.CaptureInfo()
        ));

        assertTrue(FeatureFlagManager.INSTANCE.getHasCursorVisibilityControl());
    }

    @Test
    public void hasCursorVisibilityControl_isFalseWithoutCapabilities() {
        FeatureFlagManager.INSTANCE.reset();

        assertFalse(FeatureFlagManager.INSTANCE.getHasCursorVisibilityControl());
    }

    private static void setCapabilities(PolarisCapabilities capabilities) throws Exception {
        Field field = FeatureFlagManager.class.getDeclaredField("capabilities");
        field.setAccessible(true);
        field.set(null, capabilities);
    }
}

File: app/src/test/java/com/papi/nova/manager/FeatureFlagManagerTest.kt
package com.papi.nova.manager

import com.papi.nova.api.PolarisCapabilities
import org.junit.After
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Assert.assertNull
import org.junit.Test

class FeatureFlagManagerTest {

    @After
    fun tearDown() {
        FeatureFlagManager.reset()
    }

    @Test
    fun hasCursorVisibilityControlReflectsCapabilities() {
        setCapabilities(
            PolarisCapabilities(
                server = "polaris",
                version = "1.0.0",
                features = PolarisCapabilities.Features(cursorVisibilityControl = true),
                capture = PolarisCapabilities.CaptureInfo()
            )
        )

        assertTrue(FeatureFlagManager.hasCursorVisibilityControl)
    }

    @Test
    fun hasCursorVisibilityControlIsFalseWithoutCapabilities() {
        FeatureFlagManager.reset()

        assertFalse(FeatureFlagManager.hasCursorVisibilityControl)
    }

    @Test
    fun obsoleteScopeCannotPublishOrResetNewGameCapabilities() {
        val oldScope = FeatureFlagManager.beginScope()
        val currentScope = FeatureFlagManager.beginScope()
        val current = PolarisCapabilities(
            server = "polaris",
            version = "1.3.14",
            features = PolarisCapabilities.Features(doctorV2ShadowEnabled = true),
            capture = PolarisCapabilities.CaptureInfo()
        )

        assertFalse(FeatureFlagManager.publishForScope(oldScope, current))
        assertTrue(FeatureFlagManager.publishForScope(currentScope, current))
        FeatureFlagManager.reset(oldScope)

        assertNull(FeatureFlagManager.capabilitiesForScope(oldScope))
        assertTrue(FeatureFlagManager.capabilitiesForScope(currentScope) === current)
    }

    private fun setCapabilities(capabilities: PolarisCapabilities) {
        val field = FeatureFlagManager::class.java.getDeclaredField("capabilities")
        field.isAccessible = true
        field.set(null, capabilities)
    }
}
